components/renku_data_services/solr/solr_client.py

```diff
@@ -364,3 +413,71 @@ async def delete(self, query: str) -> Response:
     async def close(self) -> None:
         """Close this client and free resources."""
         return await self.delegate.aclose()
+
+
+class SolrAdminClient(AbstractAsyncContextManager, ABC):
+    """A client to the core admin api.
+
+    Url: https://solr.apache.org/guide/solr/latest/configuration-guide/coreadmin-api.html
+    """
+
+    @abstractmethod
+    async def status(self, core_name: str | None) -> dict[str, Any] | None:
+        """Return the status of the connected core."""
+        ...
+
+    async def create(self, core_name: str | None) -> None:
+        """Create a core."""
+        ...
+
+
+class DefaultSolrAdminClient(SolrAdminClient):
+    """A client to the core admin api.
+
+    Url: https://solr.apache.org/guide/solr/latest/configuration-guide/coreadmin-api.html
+    """
+
+    delegate: AsyncClient
+    config: SolrClientConfig
+
+    def __init__(self, cfg: SolrClientConfig):
+        self.config = cfg
+        url_parsed = list(urlparse(cfg.base_url))
+        url_parsed[2] = urljoin(url_parsed[2], "/api/cores")
+        burl = urlunparse(url_parsed)
+        bauth = BasicAuth(username=cfg.user.username, password=cfg.user.password) if cfg.user is not None else None
+        self.delegate = AsyncClient(auth=bauth, base_url=burl, timeout=cfg.timeout)
+
+    async def __aenter__(self) -> Self:
+        await self.delegate.__aenter__()
+        return self
+
+    async def __aexit__(
+        self, exc_type: type[BaseException] | None, exc: BaseException | None, tb: TracebackType | None
+    ) -> None:
+        return await self.delegate.__aexit__(exc_type, exc, tb)
+
+    async def status(self, core_name: str | None) -> dict[str, Any] | None:
+        """Return the status of the connected core or the one given by `core_name`."""
+        core = core_name or self.config.core
+        resp = await self.delegate.get(f"/{core}")
+        if not resp.is_success:
+            raise SolrClientStatusException(self.config, resp)
+        else:
+            data = resp.raise_for_status().json()["status"][self.config.core]
+            # if the core doesn't exist, solr returns 200 with an empty body
+            return data if data.get("name") == self.config.core else None
+
+    async def create(self, core_name: str | None) -> None:
+        """Create a core with the given `core_name` or the name provided in the config object."""
+        core = core_name or self.config.core
+        data = {"create": {"name": core, "configSet": "_default"}}
+        resp = await self.delegate.post(
+            "",
+            content=json.dumps(data).encode("utf-8"),
+            headers={"Content-Type": "application/json"},
+        )
```

Comment:
The http client can do the json serialization and header setting for you.

```suggestion
        resp = await self.delegate.post(
            "",
            json=data,
        )
```

Reply:
Thank you!